crates/dbs-arch/src/aarch64/pmu.rs

```diff
@@ -0,0 +1,172 @@
+use thiserror::Error;
+
+/// PPI base number on aarch64.
+pub const PPI_BASE: u32 = 16;
+/// Pmu ppi number
+pub const VIRTUAL_PMU_IRQ: u32 = 7;
+
+/// Errors thrown while setting up the PMU.
+#[derive(Error, Debug)]
+pub enum PmuError {
+    /// Error while check kvm pmu capability
+    #[error("Check kvm pmu capability failed: {0}")]
+    CheckKvmPmuCap(#[source] KvmError),
+    /// Error while check pmu irq.
+    #[error("Check pmu irq error: {0}")]
+    HasPmuIrq(#[source] KvmError),
+    /// Error while check pmu init.
+    #[error("Check pmu init error: {0}")]
+    HasPmuInit(#[source] KvmError),
+    /// Error while set pmu irq.
+    #[error("Set pmu irq error: {0}")]
+    SetPmuIrq(#[source] KvmError),
+    /// Error while set pmu init.
+    #[error("Set pmu init error: {0}")]
+    SetPmuInit(#[source] KvmError),
+}
+
+type Result<T> = std::result::Result<T, PmuError>;
+
+/// Tests whether a cpu supports KVM_ARM_VCPU_PMU_V3_IRQ attribute.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn has_pmu_irq(vcpu: &VcpuFd) -> Result<()> {
+    let irq = (VIRTUAL_PMU_IRQ + PPI_BASE) as u64;
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_IRQ),
+        addr: &irq as *const u64 as u64,
+        flags: 0,
+    };
+    vcpu.has_device_attr(&attribute)
+        .map_err(PmuError::HasPmuIrq)
+}
+
+/// Tests whether a cpu supports KVM_ARM_VCPU_PMU_V3_INIT attribute.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn has_pmu_init(vcpu: &VcpuFd) -> Result<()> {
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_INIT),
+        addr: 0,
+        flags: 0,
+    };
+    vcpu.has_device_attr(&attribute)
+        .map_err(PmuError::HasPmuInit)
+}
+
+/// Set KVM_ARM_VCPU_PMU_V3_IRQ for a specific vcpu.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn set_pmu_irq(vcpu: &VcpuFd) -> Result<()> {
+    let irq = (VIRTUAL_PMU_IRQ + PPI_BASE) as u64;
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_IRQ),
+        addr: &irq as *const u64 as u64,
+        flags: 0,
+    };
+    vcpu.set_device_attr(&attribute)
+        .map_err(PmuError::SetPmuIrq)
+}
+
+/// Set KVM_ARM_VCPU_PMU_V3_INIT for a specific vcpu.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn set_pmu_init(vcpu: &VcpuFd) -> Result<()> {
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_INIT),
+        addr: 0,
+        flags: 0,
+    };
+    vcpu.set_device_attr(&attribute)
+        .map_err(PmuError::SetPmuInit)
+}
+
+/// Check kvm pmu capability
+///
+/// # Arguments
+/// * `vm` - The VM file descriptor
+fn check_kvm_pmu_cap(_vm: &VmFd) -> Result<()> {
+    // TODO: check KVM_CAP_ARM_PMU_V3 capability before setting PMU
```

Comment:
Could you create an issue about this and add issue number here as a reminder?

Reply:
I have created an issue at upstream([#215](https://github.com/rust-vmm/kvm-ioctls/issues/215)).